docs/specs/future-of-workload.md

```diff
@@ -0,0 +1,43 @@
+- Aspire.Hosting (Regular NuGet Library): This package isn't really required to be in the workload, as this is a PackageReference that all AppHost projects will have. The advantage that we get currently from having it in the workload is that this would install it into the global NuGet cache when the workload is installed, so that the AppHost projects can find it without needing to restore it from the NuGet feed. This is helpful especially for offline scenarios. The package itself provides the Aspire.Hosting library which contains all of the main logic in Aspire for building and running .NET Aspire applications.
+
+![Current state of the workload](workload-current-state.png "Current state of the workload")
+  
+## Problem statement
+
+The current design of the workload has led to some usability issues, and after a lot of deep analysis, we finally understand what is the root cause of these issues. SDK Workloads are meant to be used for tooling components, which lifetime should be tied to the lifetime of the SDK itself. When we were originally designing the workload, we believed that both DCP and Dashboard were tooling components, and that they should be installed as part of the workload. However, we have come to realize that this is not the case. DCP and Dashboard are tightly coupled to the version of Aspire.Hosting package that the AppHost project references, and therefore are part of the AppHost itself, as opposed to tooling that supports it. This is the root cause of the issues that we have seen with the workload, such as the need to ensure that the version of the workload installed matches the project reference of Aspire.Hosting, which is also a problem when trying to work with multiple Aspire projects that reference different versions of Aspire.Hosting. SDK typically makes the assumption that customers want to always be in the latest version of all of the workloads, which is likely the right thing to do for tooling components; however, this is a problem for .NET Aspire given the tight coupling between the AppHost and the DCP/Dashboard.
+
+## Proposed solution
+
+Given the above, we propose to move the DCP and Dashboard components out of the workload, and into NuGet packages that are referenced as dependencies of the Aspire.Hosting.AppHost package. This will allow the AppHost project to reference the version of Aspire.Hosting that it wants, and then the DCP and Dashboard will be installed as dependencies of that version of Aspire.Hosting. This will allow the AppHost project to have full control over the version of the DCP and Dashboard that it wants to use, and will also allow customers to have multiple AppHost projects that reference different versions of Aspire.Hosting without any issues. This will also allow us to update the DCP and Dashboard independently of the workload, which will allow us to ship updates to these components more frequently and with less risk, with a far less complex compatibility matrix.
```

Comment:
What would happen if we were to remove the `Aspire.Hosting.SDK` completely, or update it to not rely on being loaded before restore? I understand that today it does some things based on those assumptions, but I'm interested in what in the experience would change if we removed it from the workload completely. There's also the option of a NuGet-based SDK (which I've resisted so far but I think we should explore all options).

Reply:
These are the contents of `Aspire.Hosting.SDK` in question:

https://github.com/dotnet/aspire/blob/88c5c19567a12e2957e06952bbd9ef7a1a25a20e/src/Aspire.Hosting.Sdk/SDK/Sdk.targets#L9-L36

The most important part of this that **_needs_** to be evaluated before NuGet restore is the bit that is updating `ProjectReference` items. That code is what tells MSBuild that these aren't real project references, so it shouldn't treat them as such in restore. Without these, you'd have problems like restore graph of the AppHost also restoring packages coming from those project references (which could lift up dependencies, cause incompatible issues when packages are not applicable to some other project type, etc.), and it would also cause failures when those project references are targeting incompatible frameworks (as MSBuild would prevent that from working). For instance, if you had a net8.0 AppHost,  and a net9.0 frontend, that is allowed if this wasn't an SDK. @eerhardt can likely provide many more examples as he was the original one that added this logic and after different attempts found that the only way forward (if we wanted to use ProjectReference item) would be to use MSBuild SDKs.

There are alternatives though, that we can consider:
- We could decide to remove Aspire.Hosting.SDK from the workload, and instead have it be an SDK that is imported from the AppHost project itself (via `<Sdk Name="Aspire.Hosting.Sdk" Version="version" />`). The drawback of that is that it breaks existing customers as they need to change their projects, as well as MSBuild SDKs don't have a way to notify if there is an update, so people would likely get stuck on an old version of the SDK for a long time.
- We could decide not to use ProjectReference at all, and instead use a different Item for this. This would also break existing customers that would have to redefine how their apphost project references project resources, and would likely also impact some of the gestures that you get (especially in VS) automatically with Project References (e.g. drag and drop projects to automatically reference them, `dotnet add reference ..` support, etc.)

Overall, the reason why I picked this as the proposal is that the code in that SDK is very minimal and unlikely to change  going forward (proof of it being it hasn't changed since it was added), hence in favor of not breaking existing customers and not regress some of the scenarios, it feels like keeping that single SDK in the workload is a good tradeoff.

Reply:
Thanks for the extra detail. I think it's important this proposal has details of all the possible approaches and the trade-offs with each so we can make an informed decision.

> The drawback of that is that it breaks existing customers as they need to change their projects, as well as MSBuild SDKs don't have a way to notify if there is an update, so people would likely get stuck on an old version of the SDK for a long time.

I've been very resistant to use [NuGet-resolved MSBuild SDKs](https://learn.microsoft.com/visualstudio/msbuild/how-to-use-project-sdk) in the past but the reasons cited here seem like that could be reasonably mitigated with targets in `Aspire.Hosting.AppHost` that check the SDK and version being used, and if necessary emit a warning advising the user what action to take to bring them up to date. For example, assuming we make the changes in `8.2`:

- If a user updates `Aspire.Hosting.AppHost` to `8.2` but doesn't change anything else about the project, the targets within can detect that (`IsAspireHost==true`, `UsingAspireNuGetSDK!=true`) and instruct the user to remove the `IsAspireHost` property (which triggers the workload detection today) and change the project SDK to `<Project SDK="Aspire.Hosting.SDK/8.2.0">`.
- Moving forward, if a new version of `Aspire.Hosting.AppHost` introduces a requirement for a new SDK min version, that can be detected by the targets and emit a warning instructing the user to update the SDK reference to the new version.

The initial project change is one-time only, and after that a change to SDK version in the project is only required when needed (likely rarely) and the user can get clear instructions on what to do.

I'm still keen to keep using `ProjectReference` for the reasons you cited (i.e. existing tooling gestures just work) which means keeping the requirement of an MSBuild SDK so using NuGet-resolved MSBuild SDKs seems like the approach we should consider (and scrutinize) more closely.

Reply:
Just to make sure I'm following, the alternative you are suggesting is different from my proposal in that you want to remove Aspire.Hosting.SDK from the workload, and instead move it to be an SDK reference coming from AppHost project itself, correct?

Reply:
Correct. In short, I want to ensure we explore options that remove the need for the workload to be installed at all.

Reply:
Gotcha. So I did try that out, and it indeed works. I can add this to the doc as alternative solutions, but I saw two minor drawbacks with that approach:

- It makes it so that people will need that one time update on their project in order to add the import of the SDK.
- MSBuild SDKs don't have a good update mechanism, meaning if we ever wanted to ship a new update, consumers wouldn't be notified in any way, so they would have to know to go an update their csproj to use the new versions on their own. Having it on the workload, on the other hand, at least will still make it that running workload update will bring the updates, and it also means consumers will get the notification in the console and VS when their workloads have updates.

Both of these are not terrible, but given the fact that at least for the first iteration we need to still keep a workload (given we need to update users machines so DCP and Dashboard are not included in it) then I decided it was a better approach to keep the Hosting SDK in the workload and then revisit this if we look into removing the workload all together, which this proposal would totally allow.

Reply:
We can use a targetednotification (aka gold bar) in VS on Aspire projects using a certain version via whatever Kusto query gives us that info to tell them to migrate, but I am hesitant to break people (gold bars also get such a low interaction rate lol). Also in VS Code there isn't the same type of TargetedNotif infrastructure, so you would have to ask C#DK to hardcode a popup notification in there for us lol.

Reply:
I know that, at least in the past, there were some severe performance implications related to using nuget based SDKs in Visual Studio. Has that been resolved?

Reply:
If the problem you refer to is that they get restored in the UI thread, then I don't think that's fixed. However, the Hosting SDK being just a single targets file with less than 50 lines of code I'm not too concerned. (I'd be much more concerned if we were restoring Dashboard and DCP as MSBuild SDKs which at some point I tried in my prototype)

Reply:
RE this

>MSBuild SDKs don't have a good update mechanism, meaning if we ever wanted to ship a new update, consumers wouldn't be notified in any way, so they would have to know to go an update their csproj to use the new versions on their own. Having it on the workload, on the other hand, at least will still make it that running workload update will bring the updates, and it also means consumers will get the notification in the console and VS when their workloads have updates.

As I lay out in https://github.com/dotnet/aspire/pull/4708/files#r1657925144 I think we can easily improve this by adding checks in the targets brought in my `Aspire.Hosting.AppHost`.

Reply:
Yup, that's a fair point. I think that in the future if we want to not have a workload at all then having this be an MSBuild SDK would be relatively easy and we can use what you suggest to keep it up-to-date when needed. I would just rather do this change in stages, where we first focus on removing DCP and Dashboard, and then later we can move Project Templates to IDE/SDK and Hosting.SDK to be an MSBuild sdk.

Reply:
I can confirm that msbuild SDKs are still loaded on the UI thread. Jeff did a bunch of work to fix deadlocks but depending in network connectivity and size, there can be impact. msbuild sdks are fairly commonly used though and much of the perf/hang complaints come from folks with fairly complex builds and larger sdks.

Reply:
NuGet SDKs are used "a lot" (for some value of "a lot"), but I think nowhere near as much as we expect Aspire to be used.  If we move all of the large components out of the workload, then I would consider moving what's left (mainly just props and targets I think) into the .NET SDK itself rather than out into a NuGet MSBuild SDK.  You would lose the ability to version those components separately from the SDK, but I think that's actually probably an advantage, as there wouldn't be an extra, nonstandard version in the project file that had to be managed.